content/post/2024-03-12-mocking/index.Rmd

```diff
@@ -0,0 +1,143 @@
+
+This is definitely good to keep in mind whilst diving into the world of mocking.
+Not that mocking needs to be especially complex!
+
+Beside, sometimes, testing is complex because your code is begging to be refactored.
+:wink:
+
+## Revisiting the general mocking example
+
+The [example of general mocking](/2019/10/29/mocking/#general-mocking) in the post from 2019 can be rewritten to take advantage of the new mocking functionality that was [announced as no longer experimental in October 2023](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking).
+
+Now, using testthat's mocking means the best demo includes an actual package... which thankfully can be created in a few [usethis](https://usethis.r-lib.org) function calls, and writing a few code lines (well, copy-pasting them from the old post!).
+
+``` r
+create_package("../mockexample")
+# then from mockexample
+use_mit_license()
+use_r("encoding")
+use_testthat()
+use_test("encoding")
+# then filling in the script and test file
+use_git()
+use_github()
+# then remember it'd be nice to show it works
+use_github_action("check-standard")
+# Git add, commit and push
+```
+
+Find the example repository on [GitHub](https://github.com/maelle/mockexample).
+
+The mockexample package has a function called `is_encoding_a_pain()`:
+
+``` r
+is_encoding_a_pain <- function(){
+  if (Sys.info()[["sysname"]] == "Windows"){
+    return(TRUE)
+  }
+
+  FALSE
+}
+```
+
+The function returns `TRUE` on Windows, `FALSE` on other operating systems.
+
+To test it no matter the operating system, we need to make our tests believe they are run on a given operating system.
+We can in this day and age achieve this using mocking tools from testthat itself!
+Here's our test file:
+
+``` r
+test_that("is_encoding_a_pain() works on Linux", {
+  local_mocked_bindings(Sys.info = function(...) c(sysname = "Linux"))
+  expect_false(is_encoding_a_pain())
+})
+
+test_that("is_encoding_a_pain() works on Windows", {
+  local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))
+  expect_true(is_encoding_a_pain())
+})
+```
+
+The magic happens in `local_mocked_bindings()` calls such as `local_mocked_bindings(Sys.info = function(...) c(sysname = "Windows"))`, that indicates what we want the mocked function, `Sys.info()`, to return in this environment.
+Using testthat's own mocking feels good for three reasons: not needing to add another dependency (and to remember what the mocking packages are called :sweat_smile:), knowing this mocking implementation is [considered better](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking) by testthat authors, and getting a [withr vibe](https://withr.r-lib.org/) from the "local\_" aspect.
```

Comment:
Since the main motivation for this post was that mocking returns to testthat, this could be a bit more highlighted in the text, e.g., via a bulleted list

```suggestion
Using testthat's own mocking feels good for three reasons: 

- not needing to add another dependency (and to remember what the mocking packages are called :sweat_smile:)
- knowing this mocking implementation is [considered better](https://www.tidyverse.org/blog/2023/10/testthat-3-2-0/#mocking) by testthat authors, and 
- getting a [withr vibe](https://withr.r-lib.org/) from the "local\_" aspect.
```